search/includes/classes/class-search.php

```diff
@@ -2500,35 +2508,124 @@ public function load_collector(): void {
 		} );
 	}
 
-	/**
-	 * Build migration URL by replacing the host while preserving path, query, and fragment
-	 *
-	 * @param string $original_url Original URL
-	 * @param string $migration_host Migration host URL
-	 * 
-	 * @return string|null Migration URL or null if parsing fails
-	 */
-	private function build_migration_url( string $original_url, string $migration_host ): ?string {
-		$parsed_url = wp_parse_url( $original_url );
-		
+	private function build_elasticsearch_url( string $url, ?string $elasticsearch_version = null ): string {
+		if ( ! $elasticsearch_version ) {
+			// If no version is provided, we return the original URL
+			// which uses the host provided by the VIP_ELASTICSEARCH_ENDPOINTS constant
+			return $url;
+		}
+
+		$elasticsearch_seven_ports = [ 9234, 9235 ];
+		$elasticsearch_next_ports  = [ 9244, 9245 ];
+
+		$parsed_url = wp_parse_url( $url );
+
 		if ( ! $parsed_url ) {
-			return null;
+			// If parsing fails, return the original URL
+			return $url;
+		}
+
+		$host = $parsed_url['host'];
+		$port = $parsed_url['port'];
+
+		if ( $port ) {
+			if ( self::ELASTICSEARCH_MIGRATION_SEVEN === $elasticsearch_version ) {
+				// If the version is 7, replace the next version ports (when present) with the 7 ports
+				$port = str_replace( $elasticsearch_next_ports, $elasticsearch_seven_ports, $port );
+			} else {
+				// If the version is not 7, replace the 7 ports (when present) with the next version ports
+				$port = str_replace( $elasticsearch_seven_ports, $elasticsearch_next_ports, $port );
+			}
+		}
+
+		$new_url = '';
+
+		if ( isset( $parsed_url['scheme'] ) ) {
+			$new_url .= $parsed_url['scheme'] . '://';
+		}
+
+		$new_url .= $host;
+
+		if ( isset( $port ) ) {
+			$new_url .= sprintf( ':%d', $port );
 		}
 
-		$migration_url = $migration_host;
-		
 		if ( isset( $parsed_url['path'] ) ) {
-			$migration_url .= $parsed_url['path'];
+			$new_url .= $parsed_url['path'];
 		}
-		
+
 		if ( isset( $parsed_url['query'] ) ) {
-			$migration_url .= '?' . $parsed_url['query'];
+			$new_url .= '?' . $parsed_url['query'];
 		}
-		
+
 		if ( isset( $parsed_url['fragment'] ) ) {
-			$migration_url .= '#' . $parsed_url['fragment'];
+			$new_url .= '#' . $parsed_url['fragment'];
 		}
-		
-		return $migration_url;
+
+		return $new_url;
+	}
```

Comment:
Do we really need to do this url deconstruction/reconstruction?
 Could simply be:

```php
str_replace( [ ...$old_ports ], [...$new_ports ], $url );
```

Reply:
@rinatkhaziev What do you think about: https://github.com/Automattic/vip-go-mu-plugins/pull/6479/commits/8e88f2969f904a99c46d4c91a437453a44eb7cc7